### ISSUE TITLE
Quickitems

### ISSUE THREAD

**Issue by GoogleCodeExporter** [alistairreilly/andors-trail#6]
```
It would be nice to have quick access to usable items such
as potions or food. For starters, maybe a button to quickly access all
usable items?

In the future this could also be applied to temporary
skills and spells.
```

Original issue reported on code.google.com by `oskar.wi...@gmail.com` on 20 Nov 2010 at 10:01

**Comment by GoogleCodeExporter** [comment]
```
I see it being convenient for spells of skills but as far as usable items in 
general I think accessing them from the inventory is easy enough. 
```

Original comment by `Captain8...@gmail.com` on 18 Jan 2011 at 8:21

**Comment by GoogleCodeExporter** [comment]
```
I disagree. I really would like to have quicker access to food and potions.
```

Original comment by `SamuelPl...@gmail.com` on 18 Jan 2011 at 9:10

**Comment by GoogleCodeExporter** [comment]
```
How about reducing the health/experience bars to make room for such a 
quickitems button ?
The button text could be really short like "q", "Q" or "Quick".

If its going to be a single button for all quick items, then it be, or behave 
like, a drop down menu.
Click on the button to activate/show menu.
Click once on an item in the menu to use/equip it and close menu. Maybe a 
message say what you used or equipped.
Click anywhere outside the dropdown menu and the menu closes, but dont you move 
or whatever the action might be where you clicked outside.
```

Original comment by `kims...@gmail.com` on 18 Jan 2011 at 9:31

**Comment by GoogleCodeExporter** [comment]
```
I thought of 4 to 8 fully configurable slots, so I could put the meat on one, 
bonemeal on another, etc.
Could also be used for skills and spells, etc.
```

Original comment by `SamuelPl...@gmail.com` on 18 Jan 2011 at 9:35

**Comment by GoogleCodeExporter** [comment]
```
My opinion: at this state of the game there is only one action which is used 
very often: healing (aka using meat or a potion). therefore a configurable 
one-tap-button is sufficient.

now i have to tap on "hero-button", maybe tap on "inventory-tab", tap on 
meat/potion, confirm with tap on "use" and tap on "back". => 4 taps

if i am on a long run to increase experiance and i have some bad luck with 
tough monsters that is quite annoying.

a small one-tap-button with quick-heal-functionality would considerably improve 
the gameplay.

i think that the hero-button can be scaled down by a third without decreasing 
gameplay quality. reducing the health-bar provides enough place for at least 
one additional button.
```

Original comment by `erwr...@gmail.com` on 19 Jan 2011 at 5:36

**Comment by GoogleCodeExporter** [comment]
```
As I also find quick items interesting, mainly in combat mode, I started 
hacking a bit, trying to get quick items available.

I did it by adding a quick item view, that is shown when the combat view is 
visible.
It contains 3 buttons(I can add some if needed).

I plan to add the needed code to get those quick items configurable and usable.

Let me know if I should continue in that way, and if my code is welcome...

Screenshot (portrait and landscape modes): 
http://www.oleastre.be/export/andor/andor1.png
http://www.oleastre.be/export/andor/andor2.png

Current diff (I will update it when I add some functionalities):
http://www.oleastre.be/export/andor/andor.diff
```

Original comment by `oli.sa...@gmail.com` on 10 May 2011 at 10:27

**Comment by GoogleCodeExporter** [comment]
```
Again you bring another good update :) This looks good IMO. As you
stated, it would also need some way for the user to select what item
types should be in each slot. And of course, clicking the slot should
use the item (if available). Also, maybe we should have it
configurable whether to only show it during combat or always shown?

Looking forward to seeing more of your code!
/Oskar
```

Original comment by `oskar.wi...@gmail.com` on 10 May 2011 at 4:40

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Looks good! Also, I agree that healing items are very important outside of 
battolle.
```

Original comment by `jch...@email.wm.edu` on 10 May 2011 at 8:02

**Comment by GoogleCodeExporter** [comment]
```
I think it should be always shown, also a small number showing how many you 
have available would be nice, maybe have it greyed out if there is nothing 
left. Although that takes into account that only certain things will be able to 
be "quick slotted", if you can put anything in the quick slots then there would 
be no reason to have it greyed out.
```

Original comment by `sdeva...@gmail.com` on 11 May 2011 at 3:09

**Comment by GoogleCodeExporter** [comment]
```
I played a little bit more with the code and here is my current status:
- Items can be grayed if needed
- Quick items are always shown when in combat mode
- The buttons calls the ItemController when clicked

What I plan to do:
- Quick items will be available for any usable items and will be configured for 
an item type.
- Quick items will be grayed out if the count in the player's inventory is 0.

My next implementation steps:
- Implement the link between the quick items and the player inventory (quick 
item slot should listen to some kind of inventory events)
- Implements quick item for a predefined list of items
- Add user configuration support (probably using the long press menu in the 
inventory view)
- Save the user settings in the savegame
- Find a way to toggle the quick items bar display when not in combat. 
(probably add a toggle button somewhere)
```

Original comment by `oli.sa...@gmail.com` on 11 May 2011 at 7:42

**Comment by GoogleCodeExporter** [comment]
```
I forgot to add:
- display the remaining item count in the quick items bar (if I find how to do 
it ;-) )
```

Original comment by `oli.sa...@gmail.com` on 11 May 2011 at 7:44

**Comment by GoogleCodeExporter** [comment]
```
First, I think this is a great addition. Thanks!
One possible issue would be accidentally using an item. I don't know how likely 
it is; it's just something to be aware of.
```

Original comment by `jch...@email.wm.edu` on 12 May 2011 at 1:50

**Comment by GoogleCodeExporter** [comment]
```
MUST ... HAVE ... UPDATE ... :)

Screenshots are looking very nice. I hope, there is a downloadable package 
ready soon.


```

Original comment by `erwr...@gmail.com` on 12 May 2011 at 6:33

**Comment by GoogleCodeExporter** [comment]
```
Updated patch, quick items are working for a predefined set of item types.
An item is in color when it's available in the player inventory. It becomes 
grayed when everything is consumed and not available in the player inventory.

http://www.oleastre.be/export/andor/andor_2.diff

I will try to work on the configuration this week-end.

Any comment on the code is welcome.

```

Original comment by `oli.sa...@gmail.com` on 12 May 2011 at 9:41

**Comment by GoogleCodeExporter** [comment]
```
Good work, I am looking forward to this enhancement very much!
```

Original comment by `sdeva...@gmail.com` on 13 May 2011 at 4:42

**Comment by GoogleCodeExporter** [comment]
```
New version:
http://www.oleastre.be/export/andor/andor_3.diff

Done:
- Quit item works
- You can assign an usable inventory item to a quick slot via the context menu 
(long press)
- Assigned slots are saved in the game savefile.


What can be added as separate patches in the future:
- Add a shortcut to toggle the quickitem display when not in combat mode.
- Add the remaining item count on the button image.

As I added info in the savegame, old files are not valid anymore.

Let me know if the code meet your expectations...

```

Original comment by `oli.sa...@gmail.com` on 14 May 2011 at 12:07

**Comment by GoogleCodeExporter** [comment]
```
Thank you, looks great! Nice work Oli! 
This is now committed in r90. 
I added a line of "if (fileversion < 19) return;" to Inventory.java to make old 
savegames work.

Things that we could improve it with even more:
 - Maybe we should only show quickbuttons for the slots that have an assigned item.
 - Should there be some way to un-assign an item from a specific slot?

Again, nice work! Looking forward to seeing more of your code!
/Oskar
```

Original comment by `oskar.wi...@gmail.com` on 15 May 2011 at 8:49
- Added labels: **Milestone-0.6.9**

**Comment by GoogleCodeExporter** [comment]
```
Nice, I will start working on the proposed quick item enhancement.

To implement the 'un-assign' action, I can probably add a long press menu on 
each quick item button. 
That menu can have two actions: assign and un-assign.

Hiding quick buttons with no assigned item is not a problem(the detection is 
already done to change the icon). Patch will follow...
```

Original comment by `oli.sa...@gmail.com` on 16 May 2011 at 10:25

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Version 4:
http://www.oleastre.be/export/andor/andor_4.diff

- Add context menu on the quick buttons to assign/un-assign items
- Quickitem are not displayed by default in combat mode, but I added a toggle 
button in the status bar that let you display it whenever you want.
- Do not show un-assigned buttons (I think we should discuss that one, as the 
behavior seems strange with toggle button)

If you apply andor_4.diff but do not want to hide un-assigned buttons, just 
apply the following patch after andor_4.diff:
diff --git a/src/com/gpl/rpg/AndorsTrail/view/QuickitemView.java 
b/src/com/gpl/rpg/AndorsTrail/view/QuickitemView.java
index 69264a7..f94811f 100644
--- a/src/com/gpl/rpg/AndorsTrail/view/QuickitemView.java
+++ b/src/com/gpl/rpg/AndorsTrail/view/QuickitemView.java
@@ -42,7 +42,6 @@ public class QuickitemView extends FrameLayout implements 
OnClickListener{
            item.setIndex(i);
            item.setImageBitmap(world.tileStore.getBitmap(TileStore.iconID_shop));
            item.setOnClickListener(this);
-           item.setVisibility(GONE);
            item.setEmpty(true);
        }
    }
@@ -76,10 +75,8 @@ public class QuickitemView extends FrameLayout implements 
OnClickListener{
            ItemType type = world.model.player.inventory.quickitem[i];
            if(type==null){
                item.setImageBitmap(world.tileStore.getBitmap(TileStore.iconID_shop));
-               item.setVisibility(GONE);
                item.setEmpty(true);
            } else {
-               item.setVisibility(VISIBLE);
                item.setImageBitmap(world.tileStore.getBitmap(type.iconID));
                item.setEmpty(!world.model.player.inventory.hasItem(type.id));
            }

```

Original comment by `oli.sa...@gmail.com` on 17 May 2011 at 9:59

**Comment by GoogleCodeExporter** [comment]
```
Looks great. Thanks again Olivier!

Committed in r97.

I agree about the strangeness of disappearing buttons, and applied your second 
patch also :) Nice job on loading the chest-icon and switching icons when 
opening the quickbar!

This could be extended into spells/skills in the future if we implement that. 
Maybe also equipping/unequipping items.

Marking this one as ready for v0.6.9. Again, great job Olivier!
```

Original comment by `oskar.wi...@gmail.com` on 19 May 2011 at 8:00
- Changed state: **Ready**

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
I've designed a mock-up of how the quick items toolbar should look in my 
opinion.

http://wolftrak.deviantart.com/gallery/30654321

This is NOT an actual in-game screenshot.
```

Original comment by `vasilov.ivan` on 5 Jun 2011 at 4:52

**Comment by GoogleCodeExporter** [comment]
```
Mock-Up looks good. But I think that this solution is kind of an overkill. 
Basicly you have a show/hide-button, which shows up to 5 slots. The game has at 
that time only one REALY useable potion: the bone meal potion. All other 
potions are better sold then used. Actually I only need ONE slot for ONE potion 
(bone meal). In my opinion that button should be at the side of the hero 
button. The labels "HP" and "Level" could be ommited or integrated in the bars 
itself (like in other rpgs). Therefor it would be enough space for the new 
button.
```

Original comment by `erwr...@gmail.com` on 5 Jun 2011 at 9:00

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Wow, i like the look of the toolbar a lot! But instead of only using it for 
potions it should be a sort of action bar.

There are 5 useable action bar slots wich can be filled with your attack 
weappon, potions, spells, crossbow, and other usable stuff. The idea is that 
during fight the player cannot gain access to his/her inventory. So the fights 
have to be well prepared:
Do i need heal? Wich spells could be usefull at a speciffic type of monster? 
Could i use an item to blind my enemy?
This would add o lot more tactics to the later game stages. In a (hopefully) 
upcoming skilltree we could not only add skills that increase you damage/attack 
chance/health/ap points/......... , there even could be a skill that gives the 
player additional slots for his/her actionbar.

So that is my suggestion; hope you like it ;-)
```

Original comment by `michisch...@web.de` on 6 Jun 2011 at 10:50

**Comment by GoogleCodeExporter** [comment]
```
Released in v0.6.9 . Closing.
```

Original comment by `oskar.wi...@gmail.com` on 29 Jun 2011 at 7:49
- Changed state: **Done**
